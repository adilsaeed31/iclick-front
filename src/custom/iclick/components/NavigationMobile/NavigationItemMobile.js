import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import { inject, observer } from "mobx-react";
import { Router } from "routes";
import ListItemIcon from "@material-ui/core/ListItemIcon";
import ListItemText from "@material-ui/core/ListItemText";
import Divider from "@material-ui/core/Divider";
import Collapse from "@material-ui/core/Collapse";
import MenuList from "@material-ui/core/MenuList";
import ListItem from "@material-ui/core/ListItem";
import ChevronRightIcon from "mdi-material-ui/ChevronRight";
import ChevronDownIcon from "mdi-material-ui/ChevronDown";
import ChevronUpIcon from "mdi-material-ui/ChevronUp";
import { withStyles } from "@material-ui/core/styles";

const styles = (theme) => ({
  subNav: {
    marginBottom: theme.spacing.unit * 2
  },
  listItemRoot: {
    paddingTop: 16,
    paddingBottom: 16
  },
  listItemDense: {
    paddingTop: 4,
    paddingBottom: 4
  },
  listItemTextDense: {
    fontWeight: 400
  },
  listItemGutters: {
    paddingRight: 0
  },
  subMenuList: {
    paddingBottom: theme.spacing.unit * 2
  }
});

@withStyles(styles, { name: "SkNavigationItemMobile" })
@inject("routingStore", "uiStore")
@observer
class NavigationItemMobile extends Component {
  static propTypes = {
    classes: PropTypes.object,
    isTopLevel: PropTypes.bool,
    navItem: PropTypes.object,
    onClick: PropTypes.func,
    routingStore: PropTypes.object,
    shouldShowDivider: PropTypes.bool,
    uiStore: PropTypes.shape({
      closeMenuDrawer: PropTypes.func.isRequired
    })
  };

  static defaultProps = {
    classes: {},
    isTopLevel: false,
    navItem: {},
    onClick() {},
    routingStore: {},
    shouldShowDivider: true
  };

  state = { isSubNavOpen: false };

  get linkPath() {
    const { navItem, routingStore } = this.props;
    return routingStore.queryString !== ""
      ? `/tag/${navItem.slug}?${routingStore.queryString}`
      : `/tag/${navItem.slug}`;
  }

  get hasSubNavItems() {
    const { navItem: { subTags } } = this.props;
    return Array.isArray(subTags) && subTags.length > 0;
  }

  onClick = () => {
    const { navItem, uiStore, isTopLevel } = this.props;

    if (isTopLevel && this.hasSubNavItems) {
      this.props.onClick(navItem);
    } else if (this.hasSubNavItems) {
      this.setState({ isSubNavOpen: !this.state.isSubNavOpen });
    } else {
      const path = this.linkPath;
      Router.pushRoute(path, { slug: navItem.slug });
      uiStore.closeMenuDrawer();
    }
  };

  onClose = () => {
    this.setState({ isSubNavOpen: false });
  };

  renderSubNav() {
    const { classes, isTopLevel, navItem: { subTags }, uiStore, routingStore } = this.props;

    if (this.hasSubNavItems && !isTopLevel) {
      return (
        <Collapse in={this.state.isSubNavOpen} timeout="auto" unmountOnExit>
          <MenuList className={classes.subMenuList} component="div" disablePadding>
            {subTags.map(({ node: navItemGroup }, index) => (
              <NavigationItemMobile
                key={index}
                classes={classes}
                navItem={navItemGroup}
                routingStore={routingStore}
                shouldShowDivider={false}
                uiStore={uiStore}
              />
            ))}
          </MenuList>
        </Collapse>
      );
    }

    return null;
  }

  renderIcon() {
    const { classes, isTopLevel } = this.props;
    const { isSubNavOpen } = this.state;
    let icon = null;

    if (this.hasSubNavItems) {
      if (isTopLevel) {
        icon = <ChevronRightIcon />;
      } else if (isSubNavOpen) {
        icon = <ChevronUpIcon />;
      } else {
        icon = <ChevronDownIcon />;
      }
    }

    if (icon) {
      return <ListItemIcon className={classes.icon}>{icon}</ListItemIcon>;
    }

    return null;
  }

  render() {
    const { classes, navItem, shouldShowDivider } = this.props;

    return (
      <Fragment>
        <ListItem
          button
          classes={{
            root: classes.listItemRoot,
            dense: classes.listItemDense,
            gutters: classes.listItemGutters
          }}
          color="inherit"
          dense={!shouldShowDivider}
          onClick={this.onClick}
        >
          <ListItemText
            classes={{
              textDense: classes.listItemTextDense
            }}
            primary={navItem.name}
          />
          {this.renderIcon()}
        </ListItem>
        {this.renderSubNav()}
        {shouldShowDivider && <Divider />}
      </Fragment>
    );
  }
}

export default NavigationItemMobile;
