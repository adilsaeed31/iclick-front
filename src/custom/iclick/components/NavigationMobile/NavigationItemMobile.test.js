import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import NavigationItemMobile from "./NavigationItemMobile";

const mockTag = {
  _id: "cmVhY3Rpb24vdGFnOjJFcEdxYUx2TDRmNkFYRVhC",
  position: null,
  slug: "tag-slug",
  name: "Tag Name",
  subTags: []
};

const mockTagWithSubTags = {
  _id: "cmVhY3Rpb24vdGFnOjVXZE5LYXprejZ5TWdrcGY5",
  position: null,
  slug: "tag-slug",
  name: "Tag Name",
  subTags: [
    {
      node: {
        _id: "cmVhY3Rpb24vdGFnOlRFdjdSV1pDTHl1d1dmOWs2",
        position: null,
        slug: "sub-tag-slug",
        name: "Sub Tag Name",
        subTags: []
      }
    },
    {
      node: {
        _id: "cmVhY3Rpb24vdGFnOmJ2MmVEbzV3V3lvNUdNMlN3",
        position: null,
        slug: "sub-tag-slug",
        name: "Sub Tag name",
        subTags: []
      }
    }
  ]
};

const mockTagsWithMoreSubTags = {
  _id: "cmVhY3Rpb24vdGFnOkdyZ1dkYXVwOVB1ZlBYNEp6",
  position: null,
  slug: "tag-slug",
  name: "Tag Name",
  subTags: [
    {
      node: {
        _id: "cmVhY3Rpb24vdGFnOjVXZE5LYXprejZ5TWdrcGY5",
        position: null,
        slug: "sub-tag-slug",
        name: "Sub Tag Slug",
        subTags: [
          {
            node: {
              _id: "cmVhY3Rpb24vdGFnOlRFdjdSV1pDTHl1d1dmOWs2",
              position: null,
              slug: "sub-sub-tag-slug",
              name: "Sub Sub Tag name"
            }
          },
          {
            node: {
              _id: "cmVhY3Rpb24vdGFnOmJ2MmVEbzV3V3lvNUdNMlN3",
              position: null,
              slug: "sub-sub-tag-slug",
              name: "Sub Sub Tag Name"
            }
          }
        ]
      }
    },
    {
      node: {
        _id: "cmVhY3Rpb24vdGFnOkZHNnRLdkNvZGVId21Qb05p",
        position: null,
        slug: "sub-tag-slug",
        name: "Sub Tag Name",
        subTags: [
          {
            node: {
              _id: "cmVhY3Rpb24vdGFnOjREelpZdG1ydTVZaXFYSHR3",
              position: null,
              slug: "sub-sub-tag-slug",
              name: "Sub Sub Tag Name"
            }
          },
          {
            node: {
              _id: "cmVhY3Rpb24vdGFnOjg3aW1wYTNKUDVhWVFKalJE",
              position: null,
              slug: "sub-sub-tag-slug",
              name: "Sub Sub Tag Name"
            }
          }
        ]
      }
    }
  ]
};

const mockRoutingStore = {
  queryString: ""
};

const uiStore = {
  toggleMenuDrawerOpen() { },
  closeMenuDrawer() { }
};

test("renders with 1 level of tags", () => {
  const component = renderer.create(shallow(<NavigationItemMobile navItem={mockTag} routingStore={mockRoutingStore} uiStore={uiStore} />).get(3));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("renders with 2 levels of tags", () => {
  const component = renderer.create(<NavigationItemMobile navItem={mockTagWithSubTags} routingStore={mockRoutingStore} uiStore={uiStore} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("renders with 3 levels of tags", () => {
  const component = renderer.create(<NavigationItemMobile navItem={mockTagsWithMoreSubTags} routingStore={mockRoutingStore} uiStore={uiStore} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
