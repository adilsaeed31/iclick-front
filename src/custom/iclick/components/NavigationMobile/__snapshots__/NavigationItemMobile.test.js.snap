// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with 1 level of tags 1`] = `null`;

exports[`renders with 2 levels of tags 1`] = `
Array [
  <div
    className="MuiButtonBase-root-18 MuiListItem-root-7 SkNavigationItemMobile-listItemRoot-2 MuiListItem-default-10 MuiListItem-gutters-14 SkNavigationItemMobile-listItemGutters-5 MuiListItem-button-15"
    color="inherit"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="MuiListItemText-root-21"
    >
      <span
        className="MuiTypography-root-27 MuiTypography-subheading-34 MuiListItemText-primary-24"
      >
        Tag Name
      </span>
    </div>
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-54 MuiListItemIcon-root-53"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root-63"
    />
  </div>,
  <hr
    className="MuiDivider-root-74"
  />,
]
`;

exports[`renders with 3 levels of tags 1`] = `
Array [
  <div
    className="MuiButtonBase-root-18 MuiListItem-root-7 SkNavigationItemMobile-listItemRoot-2 MuiListItem-default-10 MuiListItem-gutters-14 SkNavigationItemMobile-listItemGutters-5 MuiListItem-button-15"
    color="inherit"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="MuiListItemText-root-21"
    >
      <span
        className="MuiTypography-root-27 MuiTypography-subheading-34 MuiListItemText-primary-24"
      >
        Tag Name
      </span>
    </div>
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-54 MuiListItemIcon-root-53"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root-63"
    />
  </div>,
  <hr
    className="MuiDivider-root-74"
  />,
]
`;
